linera-chain/src/manager.rs

```diff
@@ -399,33 +399,31 @@ impl ChainManager {
     pub fn create_vote(
         &mut self,
         proposal: BlockProposal,
-        outcome: BlockExecutionOutcome,
+        executed_block: ExecutedBlock,
         key_pair: Option<&KeyPair>,
         local_time: Timestamp,
+        blobs: BTreeMap<BlobId, Blob>,
     ) -> Option<Either<&Vote<ValidatedBlock>, &Vote<ConfirmedBlock>>> {
-        // Record the proposed block, so it can be supplied to clients that request it.
-        self.proposed = Some(proposal.clone());
-        self.update_current_round(local_time);
-        let ProposalContent { block, round, .. } = proposal.content;
-        let executed_block = outcome.with(block);
+        let round = proposal.content.round;
 
         // If the validated block certificate is more recent, update our locked block.
-        if let Some(lite_cert) = proposal.validated_block_certificate {
+        if let Some(lite_cert) = &proposal.validated_block_certificate {
             if self
                 .locked
                 .as_ref()
                 .map_or(true, |locked| locked.round < lite_cert.round)
             {
                 let value = Hashed::new(ValidatedBlock::new(executed_block.clone()));
-                if let Some(certificate) = lite_cert.with_value(value) {
+                if let Some(certificate) = lite_cert.clone().with_value(value) {
```

Comment:
Sorry, but I guess it could be pushed little bit further even 🤔 make `with_value(&self)` and clone the `signatures` only when creating the instance of `Certificate`. That way we don't clone unless it's actually necessary.

Reply:
I don't understand. Here we are setting the locked block to the certificate that we're creating right here. (It only returns `None` if the value doesn't match, which is unreachable.)

The place where we do throw it away is in `ChainWorkerStateWithTemporaryChanges::validate_block`. I agree this should be fixed, but it's not really related to this PR.

Reply:
That's not what I meant. I suggested to change the definition of `LiteCertificate::with_value` to work on `&self` (not `self`) and copy it after the checks.

Reply:
Ok, I didn't think of that.